source/index.ts

```diff
@@ -0,0 +1,495 @@
+		currentSegment += '\\';
+	}
+
+	switch (currentPart) {
+		case 'property': {
+			if (disallowedKeys.has(currentSegment)) {
+				return [];
+			}
+
+			parts.push(currentSegment);
+
+			break;
+		}
+
+		case 'index': {
+			throw new Error('Index was not closed');
+		}
+
+		case 'start': {
+			parts.push('');
+
+			break;
+		}
+		// No default
+	}
+
+	return parts;
+}
+
+function isStringIndex(object: unknown, key: Segment): key is string {
+	if (typeof key !== 'number' && Array.isArray(object)) {
+		const index = Number.parseInt(key, 10);
+		return Number.isInteger(index) && object[index] === object[key as unknown as number];
+	}
+
+	return false;
+}
+
+function assertNotStringIndex(object: unknown, key: Segment) {
+	if (isStringIndex(object, key)) {
+		throw new Error('Cannot use string index');
+	}
+}
+
+/**
+Get the value of the property at the given path.
+
+@param object - Object or array to get the `path` value.
+@param path - Path of the property in the object, using `.` to separate each nested key. Use `\\.` if you have a `.` in the key.
+@param defaultValue - Default value.
+
+@example
+```
+import {getProperty} from 'dot-prop';
+
+getProperty({foo: {bar: 'unicorn'}}, 'foo.bar');
+//=> 'unicorn'
+
+getProperty({foo: {bar: 'a'}}, 'foo.notDefined.deep');
+//=> undefined
+
+getProperty({foo: {bar: 'a'}}, 'foo.notDefined.deep', 'default value');
+//=> 'default value'
+
+getProperty({foo: {'dot.dot': 'unicorn'}}, 'foo.dot\\.dot');
+//=> 'unicorn'
+
+getProperty({foo: [{bar: 'unicorn'}]}, 'foo[0].bar');
+//=> 'unicorn'
+```
+*/
+export function getProperty<ObjectType, PathType extends string, DefaultValue = undefined, ReturnType = ObjectType extends AnObject | AFunction ? (unknown extends Get<ObjectType, PathType> ? DefaultValue : Get<ObjectType, PathType>) : undefined>(
+	object: ObjectType,
+	path?: PathType,
+	defaultValue?: DefaultValue,
+): ReturnType {
+	if (!isObject(object) || typeof path !== 'string') {
+		return (defaultValue === undefined ? object : defaultValue) as ReturnType;
+	}
+
+	const pathArray = getPathSegments(path);
+	if (pathArray.length === 0) {
+		return defaultValue as ReturnType;
+	}
+
+	for (let index = 0; index < pathArray.length; index++) {
+		const key = pathArray[index]!;
+
+		object = isStringIndex(object, key)
+			? (index === pathArray.length - 1 ? undefined : null) as unknown as ObjectType
+			: ((object as unknown as AnObject)[key]) as ObjectType;
+
+		if (object === undefined || object === null) {
+			// `object` is either `undefined` or `null` so we want to stop the loop, and
+			// if this is not the last bit of the path, and
+			// if it didn't return `undefined`
+			// it would return `null` if `object` is `null`
+			// but we want `get({foo: null}, 'foo.bar')` to equal `undefined`, or the supplied value, not `null`
+			if (index !== pathArray.length - 1) {
+				return defaultValue as ReturnType;
+			}
+
+			break;
+		}
+	}
+
+	return (object === undefined ? defaultValue : object) as ReturnType;
```

Comment:
Generic type `ReturnType` may need more work. Currently compatible with original typescript definition but not always accurate.